File: backend/src/merchant/api/MerchantProductApi.scala
package delivery.merchant.api

import cats.effect.IO
import delivery.merchant.objects.{UpdateProductRequest, UpdateProductResponse}
import delivery.merchant.state.MerchantDomainOps
import delivery.shared.api.ApiPlan
import delivery.shared.objects.DeliveryState
import delivery.shared.state.DeliveryStateOps
import org.typelevel.log4cats.slf4j.Slf4jLogger

object MerchantProductApi extends ApiPlan[
  MerchantProductApi.MerchantProductCommand,
  Either[String, MerchantProductApi.MerchantProductSuccess]
]:

  final case class MerchantProductCommand(
      state: DeliveryState,
      username: String,
      productId: String,
      body: UpdateProductRequest
  )

  final case class MerchantProductSuccess(nextState: DeliveryState, response: UpdateProductResponse)

  private val logger = Slf4jLogger.getLogger[IO]

  override val name: String = "MerchantProductApi"

  override def plan(input: MerchantProductCommand): IO[Either[String, MerchantProductSuccess]] =
    for
      _ <- logger.info(s"$name started, username=${input.username}, productId=${input.productId}")
      response = MerchantDomainOps.updateProduct(
        input.state.merchant,
        input.username,
        input.productId,
        input.body.name,
        input.body.description,
        input.body.price,
        input.body.remainingStock,
        input.body.listingStatus
      ).map { case (nextMerchant, updatedProduct) =>
        MerchantProductSuccess(
          DeliveryStateOps.withMerchantState(input.state, nextMerchant),
          UpdateProductResponse(ok = true, product = updatedProduct)
        )
      }
      _ <- logger.info(s"$name finished, success=${response.isRight}")
    yield response

end MerchantProductApi

File: frontend/src/apis/merchant/MerchantProductAPI.ts
import { APIMessage } from '@/apis/shared/APIMessage'
import type { TaskIO } from '@/apis/shared/TaskIO'
import { sendAPI } from '@/apis/shared/sendAPI'
import type { Product, ProductBundleGroup, ProductInventoryMode } from '@/objects/merchant/Product'
import type { UpdateProductRequest } from '@/objects/merchant/apiTypes/UpdateProductRequest'
import type { ListingStatus, ProductId } from '@/objects/shared/ids'

class MerchantProductAPI extends APIMessage<Product> {
  readonly apiName = 'merchantproductapi'
  readonly productId: ProductId
  readonly name: string
  readonly description: string
  readonly imageUrl: string
  readonly categoryName: string
  readonly price: number
  readonly remainingStock: number
  readonly listingStatus: ListingStatus
  readonly inventoryMode?: ProductInventoryMode
  readonly maxPerOrder?: number | null
  readonly bundleGroups: ProductBundleGroup[]

  constructor(
    productId: ProductId,
    name: string,
    description: string,
    imageUrl: string,
    categoryName: string,
    price: number,
    remainingStock: number,
    listingStatus: ListingStatus,
    inventoryMode?: ProductInventoryMode,
    maxPerOrder?: number | null,
    bundleGroups: ProductBundleGroup[] = [],
  ) {
    super()
    this.productId = productId
    this.name = name
    this.description = description
    this.imageUrl = imageUrl
    this.categoryName = categoryName
    this.price = price
    this.remainingStock = remainingStock
    this.listingStatus = listingStatus
    this.inventoryMode = inventoryMode
    this.maxPerOrder = maxPerOrder
    this.bundleGroups = bundleGroups
  }
}

export function updateMerchantProductIO(productId: ProductId, input: UpdateProductRequest): TaskIO<Product> {
  return sendAPI(
    new MerchantProductAPI(
      productId,
      input.name,
      input.description,
      input.imageUrl,
      input.categoryName,
      input.price,
      input.remainingStock,
      input.listingStatus,
      input.inventoryMode,
      input.maxPerOrder,
      input.bundleGroups ?? [],
    ),
  )
}
